core/src/main/java/com/linecorp/armeria/common/util/Sampler.java

```diff
@@ -42,6 +42,64 @@
  */
 @FunctionalInterface
 public interface Sampler<T> {
+
+    /**
+     * Returns a sampler that applies logical or operator to both samplers decisions.
+     */
+    default Sampler<T> or(Sampler<T> other) {
+        return new OrSampler<>(this, other);
+    }
+
+    /**
+     * Returns a sampler that applies logical and operator to both samplers decisions.
+     */
+    default Sampler<T> and(Sampler<T> other) {
+        return new AndSampler<>(this, other);
```

Comment:
```suggestion
        return new AndSampler<>(this, requireNonNull(other));
```
Can we also consider the situation where more than two samplers are chained? e.g. `a.and(b).and(c)` by accepting an array of samplers in `AndSampler.<init>()`?

Should we add a static version of `and()` and `or()`? (not sure if this is the best idea though. let me know what you think.)

Reply:
I did not like how the static function looks syntactically. Because there are no extension functions in java we need to call them such as `Sampler.and(a, b)` instead of `a.and(b)` right?

Reply:
Right. I'm fine with not adding a static method. I just wish `And/OrSampler` doesn't wrap another `And/OrSampler`, which is suboptimal.

Reply:
Hmm.. I am not sure if I am 100% following. Why wrapping another sampler is not optimal? 

If we are talking about "short circuiting" the evaluation, that's something we don't want to do.